src/class/proto_methods.rs

```diff
@@ -1,160 +1,74 @@
+use crate::ffi;
 #[cfg(not(Py_LIMITED_API))]
-use crate::class::buffer::PyBufferProcs;
-use crate::class::{
-    basic::PyObjectMethods, descr::PyDescrMethods, gc::PyGCMethods, iter::PyIterMethods,
-    mapping::PyMappingMethods, number::PyNumberMethods, pyasync::PyAsyncMethods,
-    sequence::PySequenceMethods,
-};
-use std::{
-    ptr::{self, NonNull},
-    sync::atomic::{AtomicPtr, Ordering},
-};
+use crate::ffi::PyBufferProcs;
+
+/// ABI3 doesn't have buffer APIs, so here we define the empty one.
+#[cfg(Py_LIMITED_API)]
+#[doc(hidden)]
+#[derive(Clone)]
+pub struct PyBufferProcs;
 
-/// Defines all method tables we need for object protocols.
 // Note(kngwyu): default implementations are for rust-numpy. Please don't remove them.
 pub trait PyProtoMethods {
-    fn async_methods() -> Option<NonNull<PyAsyncMethods>> {
-        None
-    }
-    fn basic_methods() -> Option<NonNull<PyObjectMethods>> {
-        None
+    fn get_type_slots() -> Vec<ffi::PyType_Slot> {
```

Comment:
> Also this would work very nicely with the buffer, where slot definitions do exist as long as it is not limited API. https://github.com/python/cpython/blob/cb115e36e1aba04b90b0ecac6f043e60064ac65b/Include/typeslots.h#L7-L8

Nice catch, thanks!